### PR TITLE
CYS: Add Zoomed Out feature

### LINKED ISSUE #50007: CYS: Zoomed out view
For the next iteration, we should introduce a zoomed-out feature in the assembler to provide a high-level view while the user is in editing mode. 

image

### PR DESCRIPTION
**Pull request by tjcafferkey**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This introduces the following for the **non-AI flow**:

* A device toolbar which allows you to switch between desktop, tablet and mobile
* A zoom out/zoom in button (this switches the device to desktop as thats the only device it works for)
* Disabled the resize handler

Closes https://github.com/woocommerce/woocommerce/issues/50007

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Ensure your CYS experience has been reset.
2. Enter the customize your store flow (non-AI).
3. Check that the device toolbar is present and clicking each device works.
4. Check that clicking zoom in/out resets the device to desktop and also zooms in/out correctly.
5. Check that the resize handler for the frame that was previously present, is no longer enabled.
6. Check that the toolbar to remove/shuffle patterns still works correctly in both zoom in/out mode.
7. Check for no regressions introduced by this PR.
8. Build and upload the plugin to an ecom plan on WP.com and check that **non of these changes are present in the AI flow**.
9. Ensure there are no regressions in the Launch Your Store flow since we updated some shared logic.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Adds zoom out mode to the CYS experience

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/components/device-toolbar.tsx [hunk 1]
```diff
@@ -0,0 +1,141 @@
+/* eslint-disable @woocommerce/dependency-group */
+/* eslint-disable @typescript-eslint/ban-ts-comment */
+/**
+ * External dependencies
+ */
+import clsx from 'clsx';
+import React from 'react';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useContext } from '@wordpress/element';
+// @ts-ignore No types for this exist yet.
```
Comment (gigitux) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/components/device-toolbar.tsx:10]: Let's use `@ts-expect-error`! In this way, we will be able to know when the type has been fixed and we will be able to remove the comment

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/hooks/use-popover-handler.ts [hunk 2]
```diff
@@ -47,7 +47,7 @@ export const usePopoverHandler = () => {
 		hoveredBlockClientId: string | null;
 	} ) => {
 		const iframe = window.document.querySelector(
-			'.woocommerce-customize-store-assembler > iframe[name="editor-canvas"]'
+			'.woocommerce-customize-store-assembler > .block-editor-iframe__container iframe[name="editor-canvas"]'
```
Comment (gigitux) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/hooks/use-popover-handler.ts:50]: Just out of curiosity, is there a particular reason for this change?

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx [hunk 3]
```diff
@@ -24,75 +28,258 @@ import { useSelect } from '@wordpress/data';
...
+		}
+
+		return () => {
+			for ( const name of eventTypes ) {
+				html.removeEventListener( name, handlers[ name ] );
+			}
+		};
+	} );
+}
+
```
Comment (gigitux) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx]: Is there a particular reason for adding this code?

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx [hunk 4]
```diff
@@ -24,75 +28,258 @@ import { useSelect } from '@wordpress/data';
...
 	const setRef = useRefEffect( ( node ) => {
 		node._load = () => {
 			setIframeDocument( node.contentDocument );
 		};
+		let iFrameDocument;
+		// Prevent the default browser action for files dropped outside of dropzones.
+		function preventFileDropDefault( event ) {
+			event.preventDefault();
+		}
```
Comment (gigitux) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx]: Is this code necessary for our project?

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/get-compatibility-styles.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+/**
+ * Returns a list of stylesheets that target the editor canvas. A stylesheet is
+ * considered targetting the editor a canvas if it contains the
+ * `editor-styles-wrapper`, `wp-block`, or `wp-block-*` class selectors.
+ *
+ * Ideally, this hook should be removed in the future and styles should be added
+ * explicitly as editor styles.
+ */
+export function getCompatibilityStyles() {
```
Comment (gigitux) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/get-compatibility-styles.js]: Is this function necessary for the zoomed-out view?

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/resizable-frame.jsx [hunk 6]
```diff
@@ -119,6 +152,18 @@ function ResizableFrame( {
...
+		}
+
+		if ( updatedWidth <= tabletStyles.width ) {
+			setDeviceType( 'Tablet' );
+		}
+
+		if ( updatedWidth <= mobileStyles.width ) {
+			setDeviceType( 'Mobile' );
+		}
+
```
Comment (gigitux) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/resizable-frame.jsx]: Given that we removed the support for `ResizableFrame,` are these changes still necessary?

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx [hunk 7]
```diff
@@ -24,75 +28,258 @@ import { useSelect } from '@wordpress/data';
...
 			delete node._load;
 			node.removeEventListener( 'load', onLoad );
+			iFrameDocument?.removeEventListener(
+				'dragover',
+				preventFileDropDefault
+			);
+			iFrameDocument?.removeEventListener(
+				'drop',
+				preventFileDropDefault
+			);
```
Comment (gigitux) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx]: Is it necessary? 🤔

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx [hunk 8]
```diff
@@ -140,11 +382,44 @@ function Iframe( {
...
+						event.nativeEvent.stopPropagation = () => {};
+						event.stopPropagation();
+						event.nativeEvent.stopPropagation = stopPropagation;
+						bubbleEvent(
+							event,
+							window.KeyboardEvent,
+							event.currentTarget
+						);
+					}
+				} }
```
Comment (gigitux) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx]: Not sure that these changes are necessary

**Review by gigitux — CHANGES_REQUESTED** [review]
Thank you for your hard work on this! The implementation is solid and shows great progress! 🚀

I have a few suggestions to consider:

1. **Code Necessity**: Ensure that all the code we added is essential for the zoomed-out view. Even if some code comes from Gutenberg, if it's not necessary, we should avoid including it to simplify future debugging.

2. **Block Toolbar Height Issue**: I noticed that the height position of the block toolbar isn't computed correctly on tablet and mobile views. Here's the link for reference:

https://github.com/user-attachments/assets/05652913-b776-4db3-bc75-703d20f44d80

Again, excellent job overall! Just a few tweaks needed before merging this PR!

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/hooks/use-popover-handler.ts (hunk 2 above)
Comment (tjcafferkey) [reply plugins/woocommerce-admin/client/customize-store/assembler-hub/hooks/use-popover-handler.ts:50]: The iframe is no longer an immediate child of `.woocommerce-customize-store-assembler` so selectors were not detecting the iframe. I tried to make it specific enough so that it would continue picking up the iframe with a level of specificity which would not result in getting the wrong iframe since there are multiple.

Comment (gigitux) [reply plugins/woocommerce-admin/client/customize-store/assembler-hub/hooks/use-popover-handler.ts:50]: It makes sense! Thanks! 🙇

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/components/device-toolbar.tsx (hunk 1 above)
Comment (tjcafferkey) [reply plugins/woocommerce-admin/client/customize-store/assembler-hub/components/device-toolbar.tsx:10]: Done in https://github.com/woocommerce/woocommerce/pull/50070/commits/df41315e0ed4ff992804284e388928baf8b8dd71

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx (hunk 3 above)
Comment (tjcafferkey) [reply plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx]: Done in https://github.com/woocommerce/woocommerce/pull/50070/commits/1015fd65e48a0f23ddd9194152e98ece8d020895

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/get-compatibility-styles.js (hunk 5 above)
Comment (tjcafferkey) [reply plugins/woocommerce-admin/client/customize-store/assembler-hub/get-compatibility-styles.js]: Done in https://github.com/woocommerce/woocommerce/pull/50070/commits/223c7c8fc9377a7d08de92bb75e9e779b7296e0d

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/resizable-frame.jsx (hunk 6 above)
Comment (tjcafferkey) [reply plugins/woocommerce-admin/client/customize-store/assembler-hub/resizable-frame.jsx]: Done in https://github.com/woocommerce/woocommerce/pull/50070/commits/aeb1c2cbb4749f37cf48423d8316b6afe2903169

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx (hunk 8 above)
Comment (tjcafferkey) [reply plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx]: Done in https://github.com/woocommerce/woocommerce/pull/50070/commits/1015fd65e48a0f23ddd9194152e98ece8d020895

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx (hunk 7 above)
Comment (tjcafferkey) [reply plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx]: Done in https://github.com/woocommerce/woocommerce/pull/50070/commits/1015fd65e48a0f23ddd9194152e98ece8d020895

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx (hunk 4 above)
Comment (tjcafferkey) [reply plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx]: Done in https://github.com/woocommerce/woocommerce/pull/50070/commits/1015fd65e48a0f23ddd9194152e98ece8d020895

**Comment by tjcafferkey** [comment]
Thanks for the review @gigitux, although I have to admit when I opened this for review at the EOD yesterday I had totally forgotten and neglected the cleanup aspect of this PR which was reflected in your comments;

>Code Necessity: Ensure that all the code we added is essential for the zoomed-out view. Even if some code comes from Gutenberg, if it's not necessary, we should avoid including it to simplify future debugging.

So, apologies as I've no doubt you spent unnecessary time trawling through this code. And thank you for the reminder!

**Comment by gigitux** [comment]
> Thanks for the review @gigitux, although I have to admit when I opened this for review at the EOD yesterday I had totally forgotten and neglected the cleanup aspect of this PR which was reflected in your comments;
> 
> > Code Necessity: Ensure that all the code we added is essential for the zoomed-out view. Even if some code comes from Gutenberg, if it's not necessary, we should avoid including it to simplify future debugging.
> 
> So, apologies as I've no doubt you spent unnecessary time trawling through this code. And thank you for the reminder!

No worries! 💪  Thanks for updating the PR 🙇

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx [hunk 9]
```diff
@@ -140,11 +266,14 @@ function Iframe( {
...
 				src={ src }
-				title={ __( 'Editor canvas', 'woocommerce' ) }
+				title={ title }
 				name="editor-canvas"
 			>
 				{ iframeDocument &&
 					createPortal(
+						// We want to prevent React events from bubbling throught the iframe
+						// we bubble these manually.
+						/* eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions */
```
Comment (gigitux) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx]: I think that we can remove this comment

**Review by gigitux — CHANGES_REQUESTED** [review]
Thanks for addressing all the feedback! I noticed two small things:

### Weird Effect When Clicking "Zoom Out" During Assembler Load

There is a strange behavior occurring when the user clicks "Zoom Out" while the assembler is still loading. You can see the issue here:

https://github.com/user-attachments/assets/39f86a7e-a3da-4e9d-86e8-dc3c855447be

We could render the topbar only when the Assembler is fully loaded. WDYT?

### Blue Border

When the icons are focused, a blue border appears around them. I think we should remove this border.

https://github.com/user-attachments/assets/f127c2d9-ee5a-4fe0-b0ae-ea20a4253121

Code-wise everything looks great! Thank you for your effort! 🙇

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx (hunk 9 above)
Comment (tjcafferkey) [reply plugins/woocommerce-admin/client/customize-store/assembler-hub/iframe.jsx]: Thanks! Fixed all of the above in:
* https://github.com/woocommerce/woocommerce/pull/50070/commits/3f4865d70e88403078ed9423975c335f374b9b52
* https://github.com/woocommerce/woocommerce/pull/50070/commits/5de9045630684d841f739c5444a80db932857a57
* https://github.com/woocommerce/woocommerce/pull/50070/commits/59537c0c4e5ce25200447e7b390ff9b389d154a6

**Review by gigitux — APPROVED** [review]
Awesome work! Let's ship it! 🚢
